bigframes/deprecation.py

```diff
@@ -0,0 +1,92 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Mark classes and functions as deprecated. """
+
+import functools
+import inspect
+import warnings
+
+
+def deprecated(reason):
+    """
+    Decorator which can be used to mark functions
+    as deprecated. It will result in a DeprecationWarning being emitted
+    when the class/function is used.
+    """
+
+    if isinstance(reason, str):
+
+        # The @deprecated is used with a 'reason'.
+        #
+        # .. code-block:: python
+        #
+        #    @deprecated("please, use another function")
+        #    def old_function(x, y):
+        #      pass
+
+        def decorator(func1):
+
+            if inspect.isclass(func1):
+                fmt1 = "Call to deprecated class {name}. {reason}"
+            else:
+                fmt1 = "Call to deprecated function {name}. {reason}"
+
+            @functools.wraps(func1)
+            def new_func1(*args, **kwargs):
+                warnings.simplefilter("always", DeprecationWarning)
```

Comment:
Please don't override people's default filters. Just use FutureWarning as recommended here: https://docs.python.org/3/library/warnings.html

See this note:

> Changed in version 3.7: Previously DeprecationWarning and FutureWarning were distinguished based on whether a feature was being removed entirely or changing its behaviour. They are now distinguished based on their intended audience and the way they’re handled by the default warnings filters.

Reply:
Done.